Nightmare Insert Archetype Player Preferences 2

BOBAMAx opened the pull request:
# About the pull request
Opened as a new PR because I managed to fuck up a merge conflict so badly I couldn't even revert the changes somehow

Changes it so that players can choose which nightmare insert role archetypes they play as instead of having them be randomly assigned. This applies to forecon as well. This also means that COs and Synth survs have dedicated slots and will not take the place of other roles in inserts.

I would say it needs a TM, but because it needs to touch every single map to add the new landmarks for co/synth in every single insert that's likely unrealistic.

# Explain why it's good for the game

Many players do not want to play as certain archetypes or have a preference for others, and should be permitted to opt in or out of them if they desire. This also allows mappers to determine exactly where co/synth survs spawn on inserts, as well as preventing co/synth survs from overriding flavor roles like CLs in inserts.


# Testing Photographs and Procedure
<details>
<summary>Screenshots & Videos</summary>

Put screenshots and videos here with an empty line between the screenshots and the `<details>` tags.

</details>


# Changelog
:cl: Drathek, BOBAMA
qol: Players can now choose to opt in or out of certain archetypes for nightmare insert survivor roles and FORECON. This is located in your player preferences.
/:cl:

BOBAMAx committed:
all of it

BOBAMAx committed:
Update survivors.dm

BOBAMAx committed:
slop

BOBAMAx committed:
Merge remote-tracking branch 'upstream/master' into nightmareprio

BOBAMAx committed:
merge

BOBAMAx committed:
Merge remote-tracking branch 'upstream/master' into yeha

BOBAMAx committed:
randomize spawn loc

BOBAMAx committed:
remove dead code

BOBAMAx committed:
Merge remote-tracking branch 'upstream/master' into yeha

BOBAMAx committed:
Merge remote-tracking branch 'upstream/master' into yeha

BOBAMAx committed:
Update clfsmugglers.dmm

BOBAMAx committed:
Fixup maps in TGM format

dbac08893db5: maps/map_files/Kutjevo/standalone/clfsmugglers.dmm

Automatically commited by: tools\mapmerge2\fixup.py

Steelpoint commented:
Mapping LGTM. Actual code testing is a separate issue.

cm13-github commented:
This pull request has conflicts, please resolve those before we can evaluate the pull request.

BOBAMAx committed:
Merge remote-tracking branch 'upstream/master' into yeha

BOBAMAx committed:
Merge remote-tracking branch 'upstream/master' into yeha

cm13-github commented:
Conflicts have been resolved. A maintainer will review the pull request shortly.

BOBAMAx committed:
Update LV522_Chances_Claim.dmm

BOBAMAx committed:
Fixup maps in TGM format

b7461b73c45b: maps/map_files/LV522_Chances_Claim/LV522_Chances_Claim.dmm

Automatically commited by: tools\mapmerge2\fixup.py

fira commented:
LGTM

BOBAMAx committed:
Merge remote-tracking branch 'upstream/master' into yeha

BOBAMAx committed:
Merge remote-tracking branch 'upstream/master' into yeha

cm13-github commented:
Conflicts have been resolved. A maintainer will review the pull request shortly.

BOBAMAx committed:
Update panic_room_hold.dmm

BOBAMAx committed:
Merge remote-tracking branch 'upstream/master' into yeha

BOBAMAx committed:
tyrargo

BOBAMAx committed:
Merge remote-tracking branch 'upstream/master' into yeha

BOBAMAx committed:
Merge remote-tracking branch 'upstream/master' into yeha

cm13-github commented:
Conflicts have been resolved. A maintainer will review the pull request shortly.

BOBAMAx committed:
synth and co prefs

BOBAMAx committed:
standard

BOBAMAx commented:
ok synths and cos can opt out of inserts and if you deselect all options as a normal surv you just dont spawn in the insert and go to back to lobby or get another job you picked (assuming theres an insert, otherwise you just spawn in)

fira commented on code/__DEFINES/__game.dm:
```diff
@@ -163,6 +163,17 @@
 #define PLAY_SURVIVOR_HOSTILE (1<<0)
 #define PLAY_SURVIVOR_NON_HOSTILE (1<<1)
 
+//toggles_insert
+#define PLAY_INSERT_STANDARD (1<<0)
+#define PLAY_INSERT_CORPORATE (1<<1)
+#define PLAY_INSERT_LEADER (1<<2)
+#define PLAY_INSERT_MEDIC (1<<3)
+#define PLAY_INSERT_ENGINEER (1<<4)
+#define PLAY_INSERT_SPECIALIST (1<<5)
+#define PLAY_INSERT_SMARTGUNNER (1<<6)
+#define PLAY_INSERT_SYNTH (1<<7)
+#define PLAY_INSERT_CO (1<<8)
```
You also need to add a DEFINE_BITFIELD clause for these, look in the code for some examples

BOBAMAx replied:
done, I think, probably.

fira commented on code/game/jobs/job/civilians/other/survivors.dm:
```diff
@@ -16,10 +16,83 @@ GLOBAL_LIST_EMPTY(spawned_survivors)
+			return TRUE
+
+	// There is no insert, so now we have to assign a random generic spawn
+	var/obj/effect/landmark/survivor_spawner/spawner = pick(generic_landmarks)
```
```suggestion
	var/obj/effect/landmark/survivor_spawner/spawner = SAFEPICK(generic_landmarks)
```

pick() crashes if there are no candidates

BOBAMAx replied:
done

fira commented on code/game/jobs/job/civilians/other/survivors.dm:
```diff
@@ -183,12 +249,48 @@ AddTimelock(/datum/job/civilian/survivor, list(
+			return TRUE
+
+	// There is no insert, so now we have to assign a random generic spawn
+	var/obj/effect/landmark/survivor_spawner/spawner = pick(generic_landmarks)
```
```suggestion
	var/obj/effect/landmark/survivor_spawner/spawner = SAFEPICK(generic_landmarks)
```
no point to check if(spawner) if it crashes

BOBAMAx replied:
done

fira commented on code/game/jobs/job/civilians/other/survivors.dm:
```diff
@@ -210,17 +312,57 @@ AddTimelock(/datum/job/civilian/survivor, list(
 /datum/job/civilian/survivor/commanding_officer/set_spawn_positions()
+	slotted_landmarks = list()
+	available_landmarks = list()
+	generic_landmarks = list()
```
That's not what set_spawn_position is for. Can you find another place to do it?

BOBAMAx replied:
split it into 2 procs and called it from where set_spawn_positions gets called in role auth

fira reviewed (changes requested):
This is missing a savfile migration so that people actually inherit the defaults, and not everything off 

Also normally we don't do TM'ing for things that involve savfile migrations, but here since you're adding something entirely new that presumably won't collide, it should be possible

cm13-github commented:
This pull request has conflicts, please resolve those before we can evaluate the pull request.

BOBAMAx committed:
stuff

BOBAMAx committed:
Merge remote-tracking branch 'upstream/master' into yeha

cm13-github commented:
Conflicts have been resolved. A maintainer will review the pull request shortly.

BOBAMAx commented:
> This is missing a savfile migration so that people actually inherit the defaults, and not everything off
> 
> Also normally we don't do TM'ing for things that involve savfile migrations, but here since you're adding something entirely new that presumably won't collide, it should be possible

I think I did it right, I nuked my local save and db and they were all enabled by default

fira commented:
Actually it's best not to TM it since it still bumps the savfile ver. But i think it's good to merge almost. The only thing missing is a map update script to update the paths you changed in map files, such as for CLF landmarks. Look in `tools/UpdatePaths/Scripts`

BOBAMAx commented:
> Actually it's best not to TM it since it still bumps the savfile ver. But i think it's good to merge almost. The only thing missing is a map update script to update the paths you changed in map files, such as for CLF landmarks. Look in `tools/UpdatePaths/Scripts`

correct me if I am mistaken but I didn't repath anything, I only added new landmarks for synths and cos, so there's nothing to replace.

fira commented:
My bad. I was thinking of the landmark changes, but these are not 1:1 replaces so they don't need a migration script.
